File: hackerrank/30 days of code/day 26, nested logic/day_26_nested_logic.py
from enum import Enum
import datetime

class HackosFine(Enum):
    HackosDaysFine = 15
    HackosMonthsFine = 500
    HackosYearsFine = 10000

class BookReturnDate:
    def __init__(self, returnedRealDate, dueDate):
        self.returnedRealDate, self.dueDate = returnedRealDate, dueDate
        self.fine = None
        self.__calculateFine()

    def __calculateFine(self):
        if self.returnedRealDate <= self.dueDate:
            self.fine = 0
        elif self.returnedRealDate.year == self.dueDate.year:
            if self.returnedRealDate.month == self.dueDate.month:
                self.fine = (self.returnedRealDate.day - self.dueDate.day) * HackosFine.HackosDaysFine.value
            else:
                self.fine = (self.returnedRealDate.month - self.dueDate.month) * HackosFine.HackosMonthsFine.value
        else:
            self.fine = HackosFine.HackosYearsFine.value
        return self.fine

    def getFine(self):
        return self.fine

def readStdinDates():
    day1, month1, year1 = map(int, input().split())
    day2, month2, year2 = map(int, input().split())

    returnedRealDate = datetime.datetime(year1, month1, day1)
    dueDate = datetime.datetime(year2, month2, day2)

    return [returnedRealDate, dueDate]

def main():
    returnedRealDate, dueDate = readStdinDates()
    bookReturnDateObj = BookReturnDate(returnedRealDate, dueDate)
    print(bookReturnDateObj.getFine())

if __name__ == "__main__":
    main()

File: hackerrank/30 days of code/day 26, nested logic/day_26_nested_logic.js
process.stdin.resume()
process.stdin.setEncoding("ascii")
_input = ""
process.stdin.on("data", function (input) {
    _input += input
})

process.stdin.on("end", function () {
   processData(_input)
})

const HackosFine = Object.freeze({
    HackosDaysFine: 15,
    HackosMonthsFine: 500,
    HackosYearsFine: 10000
})

class DateThatWorks {
    __day = 0
    __month = 0
    __year = 0

    constructor(day, month, year) {
        this.__day = day
        this.__month = month
        this.__year = year
    }

    getDay() {
        return this.__day
    }

    getMonth() {
        return this.__month
    }

    getYear() {
        return this.__year
    }
}

class BookReturnDate {
    __returnedRealDate = Object
    __dueDate = Object
    __fine = 0

    __calculateFine() {
        if (this.__returnedRealDate.getYear() < this.__dueDate.getYear())
            this.__fine = 0
        else if (this.__returnedRealDate.getYear() === this.__dueDate.getYear()) {
            if (this.__returnedRealDate.getMonth() < this.__dueDate.getMonth())
                this.__fine = 0
            else if (this.__returnedRealDate.getMonth() === this.__dueDate.getMonth()) {
                if (this.__returnedRealDate.getDay() < this.__dueDate.getDay())
                    this.__fine = 0
                else
                    this.__fine = (this.__returnedRealDate.getDay() - this.__dueDate.getDay()) * HackosFine.HackosDaysFine
            }
            else
                this.__fine = (this.__returnedRealDate.getMonth() - this.__dueDate.getMonth()) * HackosFine.HackosMonthsFine
        } else
            this.__fine = HackosFine.HackosYearsFine
    }

    constructor(returnedRealDate, dueDate) {
        this.__returnedRealDate = returnedRealDate
        this.__dueDate = dueDate
        this.__calculateFine()
    }

    getFine() {
        return this.__fine
    }
}

function readStdinDates(input) {
    const numbers = input.split(/\s+/g).map(Number)
    const returnedRealDate = new DateThatWorks(numbers[0], numbers[1], numbers[2])
    const dueDate = new DateThatWorks(numbers[3], numbers[4], numbers[5])
    return [returnedRealDate, dueDate]
}

function processData(input) {
    [returnedRealDate, dueDate] = readStdinDates(input)
    const bookReturnDateObj = new BookReturnDate(returnedRealDate, dueDate)
    console.log(bookReturnDateObj.getFine())
}
